File: HIK_Camer_Python/HIK_Camera/MvImport/HIK_Camera.py
"""
    海康工业相机 python 封装库
"""
import cv2
import sys
import numpy as np
sys.path.append("MvImport/include")
from MvImport.include.MvCameraControl_class import *

class HIK_Camera:

    def __init__(self,ExposureTime,Gain):
        # 相机参数(曝光,增益)
        self.ExposureTime = ExposureTime
        self.Gain = Gain
        # 相机参数变量
        self.nIndex = 0                                     # 相机索引
        self.camera = MvCamera()                            # 创建相机实例
        self.Camera_OK = True                               # 相机读取图片标志
        self.src = None                                     # 相机读取图片
        self.nRet = MV_OK                                   # 相机错误码
        self.deviceList = MV_CC_DEVICE_INFO_LIST()          # 枚举传输层
        self.tlayerType = MV_GIGE_DEVICE | MV_USB_DEVICE    # 设备列表
        # 相机读取图片原始数据
        self.stOutFrame = MV_FRAME_OUT()
        memset(byref(self.stOutFrame), 0, sizeof(self.stOutFrame))

    def HIKCamera_Init(self):
        # ============================= 枚举设备 =============================
        self.nRet = MvCamera.MV_CC_EnumDevices(self.tlayerType, self.deviceList)
        if self.nRet != MV_OK:
            print("MV_CC_EnumDevices fail! nRet [0x%x]" % self.nRet)
            return False

        if(self.deviceList.nDeviceNum > 0):
            for i in range(0, self.deviceList.nDeviceNum):
                mvcc_dev_info = cast(self.deviceList.pDeviceInfo[i], POINTER(MV_CC_DEVICE_INFO)).contents
                if mvcc_dev_info.nTLayerType == MV_USB_DEVICE:
                    print("HIKCamera Device: [%d]" % i)
        else:
            print("Find No HIKCamera!")
            return False

        # ========================= 选择设备并创建句柄 =========================
        stDeviceList = cast(self.deviceList.pDeviceInfo[int(self.nIndex)], POINTER(MV_CC_DEVICE_INFO)).contents       # 选择设备并创建句柄
        self.nRet = self.camera.MV_CC_CreateHandle(stDeviceList)     # 创建相机句柄
        if self.nRet != MV_OK:
            print("MV_CC_CreateHandle fail! nRet [0x%x]" % self.nRet)
            return False
        # ============================= 打开设备 =============================
        self.nRet = self.camera.MV_CC_OpenDevice(MV_ACCESS_Exclusive, 0)
        if self.nRet != MV_OK:
            print("MV_CC_OpenDevice fail! nRet [0x%x]" % self.nRet)
            return False

        return True

    def HIKCamera_startGrabbing(self):
        print("HIK_Camera open")
        # ========================= 相机开启取流 =========================
        self.nRet = self.camera.MV_CC_StartGrabbing()
        if self.nRet != MV_OK:
            print("start grabbing fail! nRet[0x%x]" % self.nRet)
            return False

        return True

    def HIKCamera_SetParam(self):
        # ========================= 设置相机曝光时间 =========================
        self.nRet = self.camera.MV_CC_SetFloatValue("ExposureTime", float(self.ExposureTime))
        if self.nRet != MV_OK:
            print("Set_ExposureTime fail! nRet[0x%x]" % self.nRet)
            return False
        # =========================== 设置相机增益 ==========================
        self.nRet = self.camera.MV_CC_SetFloatValue("Gain", float(self.Gain))
        if self.nRet != MV_OK:
            print("Set_Gain fail! nRet[0x%x]" % self.nRet)
            return False

        return True

    def HIKCamera_read(self):
        # ========================= 相机读取图像 =========================
        self.nRet = self.camera.MV_CC_GetImageBuffer(self.stOutFrame, 1000)
        if None == self.stOutFrame.pBufAddr or self.nRet != MV_OK:
            self.Camera_OK = False
            print("No data![0x%x]" % self.nRet)
        # ========================= 相机原始图像(Bayer) =========================
        pData = (c_ubyte * self.stOutFrame.stFrameInfo.nWidth * self.stOutFrame.stFrameInfo.nHeight)()
        cdll.msvcrt.memcpy(byref(pData), self.stOutFrame.pBufAddr,
                           self.stOutFrame.stFrameInfo.nWidth * self.stOutFrame.stFrameInfo.nHeight)
        data = np.frombuffer(pData, count=int(self.stOutFrame.stFrameInfo.nWidth * self.stOutFrame.stFrameInfo.nHeight),
                             dtype=np.uint8)
        image = data.reshape((self.stOutFrame.stFrameInfo.nHeight, self.stOutFrame.stFrameInfo.nWidth))
        # ========================== 相机转换图像(RGB) ==========================
        image = cv2.cvtColor(image,cv2.COLOR_BayerGB2RGB)
        self.src = image.copy()
        if (cv2.waitKey(1) == 27):
            self.Camera_OK = False
        # ========================== 相机释放图像缓存 ==========================
        self.camera.MV_CC_FreeImageBuffer(self.stOutFrame)

    def HIKCamera_close(self):
        print("HIKCamera close")
        # ========================= 相机停止取流 =========================
        self.nRet = self.camera.MV_CC_StopGrabbing()
        if self.nRet != MV_OK:
            print("stop grabbing fail! ret[0x%x]" % self.nRet)
            return False

        self.nRet = self.camera.MV_CC_CloseDevice()
        if self.nRet != MV_OK:
            print("close deivce fail! ret[0x%x]" % self.nRet)
            return False

        self.nRet = self.camera.MV_CC_DestroyHandle()
        if self.nRet != MV_OK:
            print("destroy handle fail! ret[0x%x]" % self.nRet)
            return False

        return True

File: RM_Detection/Package/HIK_Camera/src/HIK_Camera.cpp
//
// Created by steins_xin on 23-9-16.
//

#include "../include/HIK_Camera.h"

HIK_Camera::HIK_Camera() {
    //============================= 参数文件路径 =============================
    cv::FileStorage fs( PATH "HIKCamera_Param.yaml", cv::FileStorage::READ);
    //============================= 相机读取参数 =============================
    ExposureTime = (float)fs["ExposureTime"];
    Gain = (float)fs["Gain"];
    fs.release();
}

//============================= 相机使用函数 =============================

bool HIK_Camera::HIKCamera_Init() {
    //============================= 枚举设备 =============================
    nRet = MV_CC_EnumDevices(MV_GIGE_DEVICE | MV_USB_DEVICE, &stDeviceList);
    if (MV_OK != nRet) {
        printf("MV_CC_EnumDevices fail! nRet [%x]\n", nRet);
        return false;
    }
    if (stDeviceList.nDeviceNum > 0) {
        for (int i = 0; i < stDeviceList.nDeviceNum; i++) {
            printf("[HIKCamera Device %d]\n", i);
            MV_CC_DEVICE_INFO *pDeviceInfo = stDeviceList.pDeviceInfo[i];
            if (NULL == pDeviceInfo) {
                return false;
            }
        }
    } else {
        printf("Find No HIKCamera!\n");
        return false;
    }
    //========================= 选择设备并创建句柄 =========================
    nRet = MV_CC_CreateHandle(&handle, stDeviceList.pDeviceInfo[nIndex]);
    if (MV_OK != nRet) {
        printf("MV_CC_CreateHandle fail! nRet [%x]\n", nRet);
        return false;
    }
    //============================= 打开设备 =============================
    nRet = MV_CC_OpenDevice(handle);
    if (MV_OK != nRet) {
        printf("MV_CC_OpenDevice fail! nRet [%x]\n", nRet);
        return false;
    }
    return true;
}


bool HIK_Camera::HIKCamera_SetParam() {
    //========================= 设置相机曝光时间 =========================
    nRet = MV_CC_SetFloatValue(handle,"ExposureTime",ExposureTime); //调节曝光  9~10000000  MV-CU013-A0UC
    if (MV_OK != nRet) {
        printf("Set_ExposureTime fail! nRet [%x]\n", nRet);
        return false;
    }
    //=========================== 设置相机增益 ==========================
    nRet = MV_CC_SetFloatValue(handle, "Gain", Gain);  //调节增益 0~16.3704 MV-CU013-A0UC
    if (MV_OK != nRet) {
        printf("Set_Gain fail! nRet [%x]\n", nRet);
        return false;
    }
    //========================== 设置相机白平衡 ==========================
    nRet = MV_CC_SetEnumValue(handle, "BalanceWhiteAuto", 1);  //设置自动白平衡(连续) 0:Off 1:Continuous 2:Once
    if(MV_OK != nRet){
        printf("Set_BalanceWhiteAuto fail! nRet [%x]\n", nRet);
        return false;
    }
    return true;
}

bool HIK_Camera::HIKCamera_startGrabbing() {
    printf("HIK_Camera open\n");
    //========================= 相机开启取流 =========================
    nRet = MV_CC_StartGrabbing(handle);
    if (MV_OK != nRet)
    {
        printf("MV_CC_StartGrabbing fail! nRet [%x]\n", nRet);
        return false;
    }
    return true;
}

void HIK_Camera::HIKCamera_read(){
    //========================= 相机读取图像 =========================
    nRet = MV_CC_GetImageBuffer(handle, &stOutFrame, 1000);
    if (nRet == MV_OK){}
    else {
        printf("No data[0x%x]\n", nRet);
        Camera_OK = false;
    }
    //========================= 相机原始图像(Bayer) =========================
    Bayer = cv::Mat(stOutFrame.stFrameInfo.nHeight, stOutFrame.stFrameInfo.nWidth, CV_8UC1, stOutFrame.pBufAddr);
    //========================== 相机转换图像(RGB) ==========================
    cv::cvtColor(Bayer,src,cv::COLOR_BayerGB2RGB);
    if(cv::waitKey(1) == 27) Camera_OK = false;
    //========================== 相机释放图像缓存 ==========================
    MV_CC_FreeImageBuffer(handle, &stOutFrame);
}


bool HIK_Camera::HIKCamera_close() {
    printf("HIKCamera close\n");
    //========================= 相机停止取流 =========================
    nRet = MV_CC_StopGrabbing(handle);
    if (MV_OK != nRet)
    {
        printf("MV_CC_StopGrabbing fail! nRet [%x]\n", nRet);
        return false;
    }
    //========================== 关闭设备 ===========================
    nRet = MV_CC_CloseDevice(handle);
    if (MV_OK != nRet)
    {
        printf("MV_CC_CloseDevice fail! nRet [%x]\n", nRet);
        return false;
    }
    //========================== 销毁句柄 ===========================
    nRet = MV_CC_DestroyHandle(handle);
    if (MV_OK != nRet)
    {
        printf("MV_CC_DestroyHandle fail! nRet [%x]\n", nRet);
        return false;
    }
    printf("Exit\n");
    return true;
}

//============================= 相机调试函数 =============================

bool HIK_Camera::HIKCamera_outParam() {

    //========================== 相机曝光时间 ==========================
    MVCC_FLOATVALUE stExposureTime = {0};
    nRet = MV_CC_GetFloatValue(handle, "ExposureTime", &stExposureTime);
    if (MV_OK == nRet)
    {
        printf("ExposureTime current value:%f\n", stExposureTime.fCurValue);   //当前曝光
        printf("ExposureTime max value:%f\n", stExposureTime.fMax);           //最大设置曝光
        printf("ExposureTime min value:%f\n\n", stExposureTime.fMin);        //最小设置曝光
    }
    else
    {
        printf("Get ExposureTime failed! nRet [%x]\n\n", nRet);
        return false;
    }

    //=========================== 相机增益 ==========================
    MVCC_FLOATVALUE GainValue = {0};
    nRet = MV_CC_GetFloatValue(handle, "Gain", &GainValue);
    if (MV_OK == nRet)
    {
        printf("Gain current value:%f\n", GainValue.fCurValue);   //当前增益
        printf("Gain max value:%f\n", GainValue.fMax);           //最大设置增益
        printf("Gain min value:%f\n\n", GainValue.fMin);        //最小设置增益
    }
    else
    {
        printf("Get Gain failed! nRet [%x]\n\n", nRet);
        return false;
    }

    return true;
}

bool HIK_Camera::HIKCamera_FPS() {

    //=========================== 相机帧率 ==========================
    MVCC_FLOATVALUE FPS = {0};
    nRet = MV_CC_GetFloatValue(handle, "ResultingFrameRate", &FPS);
    if (MV_OK == nRet)
    {
        printf("FPS current value:%f\n", FPS.fCurValue);   //当前帧率
    }
    else
    {
        printf("FPS failed! nRet [%x]\n\n", nRet);
        return false;
    }

    return true;
}




